server.c

```diff
@@ -0,0 +1,264 @@
+    }
+
+  int size;
+  if (recv_wrapper (socket_fd, &size, sizeof (int), 0) == S_FAILURE)
+    {
+      print_error ("Could not receive data from client\n");
+      return (S_FAILURE);
+    }
+
+  if (size != 0)
+    {
+      if (recv_wrapper (socket_fd, password, sizeof (password_t), 0)
+          == S_FAILURE)
+        {
+          print_error ("Could not receive data from client\n");
+          return (S_FAILURE);
+        }
+    }
+
+  return (S_SUCCESS);
+}
+
+static void *
+handle_client (void *arg)
+{
+  cl_context_t *cl_ctx = (cl_context_t *)arg;
+  cl_context_t local_ctx = *cl_ctx;
+  mt_context_t *mt_ctx = &local_ctx.context->context;
+
+  if (pthread_mutex_unlock (&cl_ctx->context->context.mutex) != 0)
+    {
+      print_error ("Could not unlock mutex\n");
+      return (NULL);
+    }
+
+  if (send_wrapper (local_ctx.socket_fd, mt_ctx->config->hash, HASH_LENGTH, 0)
+      == S_FAILURE)
+    {
+      print_error ("Could not send hash to client\n");
+      return (NULL);
+    }
+
+  while (true)
+    {
+      task_t task;
+      if (queue_pop (&mt_ctx->queue, &task) == S_FAILURE)
+        return (NULL);
+
+      task.to = task.from;
+      task.from = 0;
+
+      if (delegate_task (local_ctx.socket_fd, &task, mt_ctx->password)
+          == S_FAILURE)
+        {
+          if (queue_push (&mt_ctx->queue, &task) == S_FAILURE)
+            print_error ("Could not push to the queue\n");
+          return (NULL);
+        }
+
+      if (pthread_mutex_lock (&mt_ctx->mutex) != 0)
+        {
+          print_error ("Could not lock a mutex\n");
+          return (NULL);
+        }
+      pthread_cleanup_push (cleanup_mutex_unlock, &mt_ctx->mutex);
+
+      --mt_ctx->passwords_remaining;
+      if (mt_ctx->passwords_remaining == 0 || mt_ctx->password[0] != 0)
+        if (pthread_cond_signal (&mt_ctx->cond_sem) != 0)
+          {
+            print_error ("Could not signal a condition\n");
+            return (NULL);
+          }
+
+      pthread_cleanup_pop (!0);
+    }
+
+  close (local_ctx.socket_fd);
+
+  return (NULL);
+}
+
+static void *
+handle_clients (void *arg)
+{
+  serv_context_t *serv_ctx = (serv_context_t *)arg;
+  mt_context_t *mt_ctx = &serv_ctx->context;
+
+  while (true)
+    {
+      int client_socket = accept (serv_ctx->socket_fd, NULL, NULL);
+      if (client_socket == -1)
+        {
+          print_error ("Could not accept new connection\n");
+          continue;
+        }
+
+      cl_context_t cl_ctx = {
+        .context = serv_ctx,
+        .socket_fd = client_socket,
+      };
+
+      if (thread_create (&mt_ctx->thread_pool, handle_client, &cl_ctx)
+          == S_FAILURE)
+        {
+          print_error ("Could not create client thread\n");
+          if (pthread_mutex_unlock (&mt_ctx->mutex) != 0)
+            {
+              print_error ("Could not lock mutex\n");
+              return (NULL);
+            }
+          continue;
+        }
+
+      if (pthread_mutex_lock (&mt_ctx->mutex) != 0)
```

Comment:
Doesn't look sane. There should be one more lock before thread_create.

Reply:
There is one more lock after thread_create. I tried to add one more lock after this lock and program froze.